File: web/src/TrackEditForm.ts
import { Track } from "./Library";
import { TrackUpdate } from "./generated/messages";
import { FormatPlaybackPositionWithMillis } from "./PlaybackPositionFormatters";

// ratings are stored 0-100 but shown as five stars in half steps, like itunes
export const RATING_PER_STAR = 20;

// years are pushed back into itunes as an int32, so anything larger is rejected
const MAX_INT32 = 2147483647;

// m:ss.mmm parser mirroring the ios edit form: strict minutes:seconds with up
// to three optional millis digits, returning null when it doesn't match
export function parsePlaybackTime(value: string): number | null {
  const match = value.match(/^(\d+):([0-5]\d)(?:\.(\d{0,3}))?$/);
  if (!match) {
    return null;
  }
  const minutes = Number(match[1]);
  const seconds = Number(match[2]);
  const fraction = match[3] ? Number(`0.${match[3]}`) : 0;
  return minutes * 60 + seconds + fraction;
}

// the edit form's working copy of a track's editable fields, kept as the
// strings being typed; mirrors the ios edit form: only changed fields are
// submitted and sort names go stale until the next sync
export interface TrackEditForm {
  name: string;
  artist: string;
  album: string;
  albumArtist: string;
  genre: string;
  year: string;
  start: string;
  finish: string;
  // in stars, 0 to 5 in half steps
  rating: number;
}

export function formFromTrack(track: Track): TrackEditForm {
  return {
    name: track.name,
    artist: track.artistName,
    album: track.albumName,
    albumArtist: track.albumArtistName,
    genre: track.genre,
    year: String(track.year),
    start: FormatPlaybackPositionWithMillis(track.start),
    finish: FormatPlaybackPositionWithMillis(track.finish),
    rating: track.rating / RATING_PER_STAR,
  };
}

export const isNameValid = (form: TrackEditForm) => form.name !== "";
export const isArtistValid = (form: TrackEditForm) => form.artist !== "";
export const isGenreValid = (form: TrackEditForm) => form.genre !== "";
export const isYearValid = (form: TrackEditForm) =>
  /^\d+$/.test(form.year) && Number(form.year) <= MAX_INT32;

export const isStartValid = (form: TrackEditForm, duration: number) =>
  isPositionValid(form.start, duration);
export const isFinishValid = (form: TrackEditForm, duration: number) =>
  isPositionValid(form.finish, duration);

function isPositionValid(value: string, duration: number): boolean {
  const seconds = parsePlaybackTime(value);
  return seconds !== null && seconds < duration + 0.0005;
}

export function isFormValid(form: TrackEditForm, duration: number): boolean {
  return (
    isNameValid(form) &&
    isArtistValid(form) &&
    isGenreValid(form) &&
    isYearValid(form) &&
    isStartValid(form, duration) &&
    isFinishValid(form, duration)
  );
}

function ratingValue(form: TrackEditForm): number {
  return Math.round(form.rating * RATING_PER_STAR);
}

// a track update carrying only the fields that differ from the track, matching
// the ios edit form; start & finish are sent as seconds
export function changedFields(form: TrackEditForm, track: Track): TrackUpdate {
  const update = new TrackUpdate();
  if (form.name !== track.name) {
    update.name = form.name;
  }
  if (form.artist !== track.artistName) {
    update.artist = form.artist;
  }
  if (form.album !== track.albumName) {
    update.album = form.album;
  }
  if (form.albumArtist !== track.albumArtistName) {
    update.albumArtist = form.albumArtist;
  }
  if (form.genre !== track.genre) {
    update.genre = form.genre;
  }
  if (form.year !== String(track.year) && /^\d+$/.test(form.year)) {
    update.year = Number(form.year);
  }
  const start = parsePlaybackTime(form.start);
  if (
    form.start !== FormatPlaybackPositionWithMillis(track.start) &&
    start !== null
  ) {
    update.start = start;
  }
  const finish = parsePlaybackTime(form.finish);
  if (
    form.finish !== FormatPlaybackPositionWithMillis(track.finish) &&
    finish !== null
  ) {
    update.finish = finish;
  }
  if (ratingValue(form) !== track.rating) {
    update.rating = ratingValue(form);
  }
  return update;
}

// whether an update carries any field, so an unchanged save can skip the push
export function hasChanges(update: TrackUpdate): boolean {
  return (
    update.has_name ||
    update.has_artist ||
    update.has_album ||
    update.has_albumArtist ||
    update.has_genre ||
    update.has_year ||
    update.has_start ||
    update.has_finish ||
    update.has_rating
  );
}

// a copy of the track with the edits applied; sort names are left as they were
// until the next sync, matching the ios edit form
export function updatedTrack(form: TrackEditForm, track: Track): Track {
  return {
    ...track,
    name: form.name,
    artistName: form.artist,
    albumName: form.album,
    albumArtistName: form.albumArtist,
    genre: form.genre,
    year: /^\d+$/.test(form.year) ? Number(form.year) : track.year,
    start: parsePlaybackTime(form.start) ?? track.start,
    finish: parsePlaybackTime(form.finish) ?? track.finish,
    rating: ratingValue(form),
  };
}

File: ios/Warehouse/TrackEditForm.swift
import Foundation

/// m:ss.mmm helpers for the start & finish fields, mirroring the web app's
/// playback position formatters; trailing zeros in the millis are trimmed
enum PlaybackTimeMillis {
    static func format(_ time: TimeInterval) -> String {
        var totalMillis = Int((time * 1000).rounded())
        if totalMillis < 0 { totalMillis = 0 }
        let minutes = totalMillis / 60_000
        let seconds = totalMillis / 1000 % 60
        let millis = totalMillis % 1000
        let base = String(format: "%d:%02d", minutes, seconds)
        guard millis > 0 else { return base }
        var trimmed = String(format: "%03d", millis)
        while trimmed.hasSuffix("0") {
            trimmed.removeLast()
        }
        return "\(base).\(trimmed)"
    }

    static func parse(_ value: String) -> TimeInterval? {
        guard let match = value.wholeMatch(of: #/([0-9]+):([0-5][0-9])(?:\.([0-9]{0,3}))?/#) else {
            return nil
        }
        guard let minutes = Double(match.1), let seconds = Double(match.2) else { return nil }
        var fraction = 0.0
        if let digits = match.3, !digits.isEmpty {
            fraction = Double("0.\(digits)") ?? 0
        }
        return minutes * 60 + seconds + fraction
    }
}

/// the edit sheet's working copy of a track's editable fields, kept as the
/// strings being typed; mirrors the web app's edit track panel semantics:
/// only changed fields are submitted & sort names go stale until the next sync
struct TrackEditForm: Equatable {
    /// ratings are stored 0-100 but shown as 5 stars, like itunes
    static let ratingMultiplier = 20.0

    var name: String
    var artist: String
    var album: String
    var albumArtist: String
    var genre: String
    var year: String
    var start: String
    var finish: String
    /// in stars, 0 to 5 in half steps
    var rating: Double
    var artworkFilename: String?
    var artworkCleared = false

    init(song: Song) {
        name = song.name
        artist = song.artistName
        album = song.albumName
        albumArtist = song.albumArtistName
        genre = song.genre
        year = String(song.year)
        start = PlaybackTimeMillis.format(song.start)
        finish = PlaybackTimeMillis.format(song.finish)
        rating = Double(song.rating) / Self.ratingMultiplier
        artworkFilename = song.artworkFilename
    }

    var isNameValid: Bool { !name.isEmpty }
    var isArtistValid: Bool { !artist.isEmpty }
    var isGenreValid: Bool { !genre.isEmpty }

    var isYearValid: Bool {
        year.wholeMatch(of: #/[0-9]+/#) != nil && Int32(year) != nil
    }

    func isStartValid(duration: TimeInterval) -> Bool {
        isValidPosition(start, duration: duration)
    }

    func isFinishValid(duration: TimeInterval) -> Bool {
        isValidPosition(finish, duration: duration)
    }

    func isValid(duration: TimeInterval) -> Bool {
        isNameValid && isArtistValid && isGenreValid && isYearValid
            && isStartValid(duration: duration) && isFinishValid(duration: duration)
    }

    /// a track update with only the fields that differ from the song set;
    /// start & finish are sent as seconds
    func changedFields(from song: Song) -> TrackUpdate {
        var update = TrackUpdate()
        if name != song.name { update.name = name }
        if artist != song.artistName { update.artist = artist }
        if album != song.albumName { update.album = album }
        if albumArtist != song.albumArtistName { update.albumArtist = albumArtist }
        if genre != song.genre { update.genre = genre }
        if year != String(song.year), let year = Int32(year) { update.year = year }
        if start != PlaybackTimeMillis.format(song.start), let seconds = PlaybackTimeMillis.parse(start) {
            update.start = seconds
        }
        if finish != PlaybackTimeMillis.format(song.finish), let seconds = PlaybackTimeMillis.parse(finish) {
            update.finish = seconds
        }
        if ratingValue != song.rating { update.rating = Int32(ratingValue) }
        if artworkCleared {
            if song.artworkFilename != nil { update.artwork = "" }
        } else if let artworkFilename, artworkFilename != song.artworkFilename {
            update.artwork = artworkFilename
        }
        return update
    }

    /// a copy of the song with the edits applied; sort names are left as they
    /// were until the next sync, matching the web app
    func updatedSong(from song: Song) -> Song {
        Song(
            id: song.id,
            name: name,
            sortName: song.sortName,
            artistName: artist,
            artistSortName: song.artistSortName,
            albumArtistName: albumArtist,
            albumArtistSortName: song.albumArtistSortName,
            albumName: album,
            albumSortName: song.albumSortName,
            genre: genre,
            year: Int(year) ?? song.year,
            duration: song.duration,
            start: PlaybackTimeMillis.parse(start) ?? song.start,
            finish: PlaybackTimeMillis.parse(finish) ?? song.finish,
            discNumber: song.discNumber,
            trackNumber: song.trackNumber,
            playCount: song.playCount,
            rating: ratingValue,
            musicFilename: song.musicFilename,
            artworkFilename: artworkCleared ? nil : artworkFilename,
            addedDate: song.addedDate)
    }

    /// the stars back in 0-100 form
    private var ratingValue: Int {
        Int((rating * Self.ratingMultiplier).rounded())
    }

    private func isValidPosition(_ value: String, duration: TimeInterval) -> Bool {
        guard let seconds = PlaybackTimeMillis.parse(value) else { return false }
        return seconds < duration + 0.0005
    }
}
